packages/quantification/src/net-quantification.ts

```diff
@@ -0,0 +1,185 @@
+ * 1. Years with negative numbers are set to zero
+ * 2. The amount emitted will be debited from the subsequent years of the same field
+ * 3. Should the amount emitted exceed the amount available from subsequent years, wrap to the
+ * start year and continue to the currently considered year
+ * 4. Should the amount exceed all available tonnes removed from this project, repeat step 1
+ * starting with the currently considered year in the next quantification
+ * 5. Should the negative amount exceed the total tonnes available, the negative amount should
+ * persist in the year it originated
+ *
+ * An example:
+ *
+ * Starting with:
+ * | field | 2015 | 2016 | 2017 |
+ * |-------|------|------|------|
+ * | A     | 10   | -30  | 9    |
+ * | B     | 10   | 10   | 9    |
+ *
+ * 30 left
+ * | field | 2015 | 2016 | 2017 |
+ * |-------|------|------|------|
+ * | A     | 10   | 0    | 9    |
+ * | B     | 10   | 10   | 9    |
+ *
+ * 21 left
+ * | field | 2015 | 2016 | 2017 |
+ * |-------|------|------|------|
+ * | A     | 10   | 0    | 0    |
+ * | B     | 10   | 10   | 9    |
+ *
+ * 11 left
+ * | field | 2015 | 2016 | 2017 |
+ * |-------|------|------|------|
+ * | A     | 0    | 0    | 0    |
+ * | B     | 10   | 10   | 9    |
+ *
+ * 1 left
+ * | field | 2015 | 2016 | 2017 |
+ * |-------|------|------|------|
+ * | A     | 0    | 0    | 0    |
+ * | B     | 10   | 0    | 9    |
+ *
+ * 0
+ * | field | 2015 | 2016 | 2017 |
+ * |-------|------|------|------|
+ * | A     | 0    | 0    | 0    |
+ * | B     | 10   | 0    | 8    |
+ * | final | 10   | 0    | 8    |
+ *
+ */
+export const getNetQuantificationProjection = (
+  quantifications: Pick<
+    UnadjustedQuantificationSummary,
+    'somscAnnualDifferencesBetweenFutureAndBaselineScenarios'
+  >[],
+  logger?: Pick<Console, 'debug' | 'table'>
+): AnnualTotals[] => {
+  const netQuantifications: AnnualTotals[] = quantifications.map(
+    (quantification) => ({
+      ...quantification.somscAnnualDifferencesBetweenFutureAndBaselineScenarios,
+    })
+  );
+
+  // Get the set of years in all quantifications
+  const years = new Set<string>();
+  for (const quantification of netQuantifications) {
+    for (const year of Object.keys(quantification)) {
+      years.add(year);
+    }
+  }
+
+  /**
+   * CAUTION -- years are represented as strings and sorted alphabetically. If for some reason, we
+   * begin supporting years < 1000 or > 9999, we'll need to rework this sort
+   */
+  const yearsOrderedAsc = Array.from(years.values()).sort();
+  logger?.debug('All years:', yearsOrderedAsc);
+
+  // No guarantee that all fields have data for each year, so we need to add zeroes for the missing years
+  for (const quantification of netQuantifications) {
+    for (const year of years) {
+      if (quantification[year] === undefined) {
+        quantification[year] = 0;
+      }
+    }
+  }
+
+  let rowIndex = 0;
+  let colIndex = 0;
+  let debt = 0;
+  while (rowIndex < netQuantifications.length) {
```

Comment:
Generally we have tried to avoid writing imperative loops in favor of functional programming methods like `map`, `reduce`, `forEach`, etc... And so this to me seems like it would be better suited for a reducer (for the sake of keeping track of `debt` as the accumulator), or as a simple `forEach`.

At the same time, I can't say I'm too opinionated on the imperative vs. functional programming, although seeing while loops never fails to make me feel nervous (I've written too many infinite loops in my time).

So curious about your take on this debate Dave, and also @jaycenhorton. We should just decide whether or not we want to choose to be definitively opinionated about this, or say just go with whatever helps you write your code and don't worry about functional vs. imperative.

Reply:
I'm in favor of whatever is most readable ("Code is for humans" is the guiding principle I've heard). If we think using something functional would help here, I'll change it. Writing it with the while loops was the easiest thing for me to do.

Reply:
Here's an attempt at cooking it up in a more functional style (this definitely still has a bug when the list needs to wrap back to the beginning of the fields if there aren't enough)

```ts
import * as math from 'mathjs';

const fields = [[5, -1, 2, -5, 1, 0, 1], [-5, 1, -5], [8]];

const reduceDebtForField = (debt) => (vintages, vintage, index) => {
  debt = math.add(Math.min(vintage, 0), debt);
  vintages[index] = math.abs(debt) >= vintage ? 0 : math.add(vintage, debt);
  if (debt && vintages[index + 1]) {
    vintages[index + 1] = math.add(vintages[index + 1], debt);
    debt = 0;
  } else if (debt) {
    debt = Math.min(0, vintage);
  }
  if (index === vintages.length - 1 && debt < 0 && math.sum(vintages) > 0) {
    vintages = vintages.reduce(reduceDebtForField(debt), vintages);
  }
  return vintages;
};

const reduceDebtForProject = (accumulator, next, index) => {
  let field = next;
  const fieldSum = math.sum(field);
  let debt = math.add(accumulator.debt, Math.min(fieldSum, 0));
  if (fieldSum <= 0) {
    field = math.zeros(math.size(field) as number[]);
  } else {
    field = field.reduce(reduceDebtForField(debt), field);
    debt = Math.min(0, math.sum(field));
  }
  if (index === fields.length - 1 && accumulator.debt + fieldSum < 0) {
    accumulator.fields.reduce(
      reduceDebtForProject,
      { debt: accumulator.debt + fieldSum, fields: accumulator.fields },
      []
    );
  }
  accumulator.fields = [...accumulator.fields, field];
  accumulator.debt = debt;
  return accumulator;
};

const calculated = fields.reduce(reduceDebtForProject, { debt: 0, fields: [] });

console.log('calculated:', calculated);
```

Note that this works for most cases, but when you set the last array to a number <8 it starts to break due to a recursion bug in my logic somewhere

Reply:
Do you feel this makes the code more readable versus the proposed algorithm? How much do we value the implementation paradigm versus readability?

Happy to save this for an engineering time. Let me know if you'd like me to incorporate this functional style in to the algorithm and we can compare the two.

Reply:
I think your approach is sufficient but I don't find either (yours nor mine) readable tbh. I think a combination of the two (dropping while loops and adding in the mathjs usage) would be a huge improvement.

Reply:
Short term, i'm comfortable with merging either variant (assuming the above had it's bugs kinked out)